api/types/role.go

```diff
@@ -588,28 +574,6 @@ func (r *Rule) CheckAndSetDefaults() error {
 	if len(r.Verbs) == 0 {
 		return trace.BadParameter("missing verbs")
 	}
-	if len(r.Where) != 0 {
```

Comment:
I'm not sure removing these checks here and above is what we want. Potentially opens the way for incorrectly formatted roles/rules in the database and if they're parsed elsewhere, spreads out the validation logic across the codebase. Any way to keep these? Maybe move parsing-related code in a separate package that the client can import.

Reply:
Should we move this logic to the server side to reduce the size and scope of the client, in `auth/apiserver.upsertRole`? Just made a commit showing this change.

This would also be a nice change for when we implement the client in other languages, so that we don't have to translate much logic that has to be maintained. I'm hoping we can get to a place where other client libraries can be auto-generated, or at least quickly translated without much work. 

Maybe all `CheckAndSetDefaults` functions could be moved to services, to be called by the server, to reduce the size/scope of the client. Same with some other functions that have been moved to types, such as rule.score and rule.MatchesWhere. Part of my work now is detangling the api from our inner services, so I'm trying to move some of these methods to exported functions in services.

@russjones @awly @r0mant Bit of a brain dump but what do you guys think?

Reply:
This also applies to the `Check` method on `CertAuthority` which I'm refactoring right now. 

https://github.com/gravitational/teleport/blob/master/api/types/authority.go#L483-L502

Reply:
@r0mant @awly Like @Joerger mentioned, we need to move rule parsing out of the client otherwise we'll have to implement the same parser in every language we create API bindings for. For example, Python is next.

The plan is to do basic validation of the struct client side (are fields missing that should be set) then doing the more involved logic (like running the parser) server side before upserting the role.

What do you think?

Reply:
Server-side parsing sounds good, as long as we detect problems at role creation time vs evaluation time.
Will the check be moved to server-side validation in this PR?

Reply:
Yes, I added it to `apiserver.go`

Reply:
@awly @r0mant I chose to split up the validation logic for `CertAuthority` in the same way in #5295. The goal of this is to keep basic validation in `CheckAndSetDefaults` and separate any validation that requires real logic/dependencies in services.

I've talked with @russjones about this, and the plan is to do the least movement/refactoring necessary to reduce dependencies for now to reach MVP, and come back to this in a future PR. If you guys think this is a problem and shouldn't be merged as such, we can consider upping the priority of refactoring validation to the server side.

Likewise, there is a lot of other code to trim out of the types package, such as Certificate Rotation logic, which we'll need to come back to later in the API refactoring.